### ISSUE TITLE
Feature request: Stop entire test suite execution on beforeAll failure

### ISSUE THREAD

**Issue by juliemr** [jasmine/jasmine#1533]
This is somewhat related to issue #577, but different because it deals with beforeAll instead of beforeEach. In many situations, I imagine that a failure in beforeAll means that there's no useful information to be gained from continuing the test. It would be nice to immediately fail without trying to run each of the specs after that.

## Expected Behavior
```js
describe('bad setup', function() {
  beforeAll(function() {
    throw new Error('error from beforeAll');
  });

  it('test 1', function() {
    expect(1).toEqual(11);
  });
});
```

Expected behavior - output:
Suite error: bad setup
  Message:
    Error: error from beforeAll

## Current Behavior
Output tries to run all tests:
Failures:
1) bad setup test 1
  Message:
    Expected 1 to equal 11.
  Stack:
    ...

2) bad setup test 2
  Message:
    Expected 2 to equal 22.
  Stack:
    ...

Suite error: bad error
  Message:
    Error: error from beforeAll

## Suite that reproduces the behavior (for bugs)
```javascript
describe('bad setup', function() {
  beforeAll(function() {
    throw new Error('error from beforeAll');
  });

  it('test 1', function() {
    expect(1).toEqual(11);
  });
});
```

## Your Environment
Jasmine 3.1. I believe this behavior is the same in all environments. I have tried both with and without `stopSpecOnExpectationFailure`

**Comment by slackersoft** [comment]
Does it make sense to merge this into the existing option to stop a spec on expectation failure or does this make sense as a separate option? The option to stop the suite execution (`failFast`) should do this as well, but it obviously has other effects as well.

**Comment by juliemr** [comment]
My two cents is that it makes sense to merge with the existing `stopSpecOnExpectationFailure` option.

**Comment by slackersoft** [comment]
That makes sense to me. I'd be happy to review a pull request for this.

**Comment by codehunter13** [comment]
any progress on this :)

**Comment by Durgam777** [comment]
Any update on this, please

Thanks,

**Comment by MicahZoltu** [comment]
It is unclear to me why specs are executed after `beforeAll` fails.  What is the use-case for a `beforeAll` where you are OK with it failing?  All of the use cases for `beforeAll` I can think of are for doing setup work that is necessary before the system is in a reasonable state to run any tests.  For example, if you need to spin up an in-memory database that your tests reference (an expensive operation that you want to do once), if that DB never starts up none of the tests will pass because there is nothing for them to test against.

In other words, I can't think of any reason why this needs to be an option.  Who would turn the option off?  Why would someone turn the option off?  What sort of `beforeAll` operations are both reasonable and optional for test execution?

**Comment by slackersoft** [comment]
Jasmine has always attempted to show the user all of the expectation failures that occur during the execution of a test. In order to do this, a single failing expectation can't always cause even the current function (`it`, `beforeEach`, `afterEach`, etc) to stop running.

I would _still_ be happy to review a pull request to make a `beforeAll` act similar to a `beforeEach` when the `oneFailurePerSpec` option is set. This functionality would mean that with this option set a failure in a `beforeAll` would cause execution to proceed to the first `afterAll` for that same suite without executing any remaining `beforeAll`s or `it`s.

**Comment by DenysVuika** [comment]
Any progress on this?

This feature may save loads of time and money when using with CI/CD and large amounts of e2e tests.

**Comment by stevenmchaves** [comment]
Interested in this resolution as well.

**Comment by conartist6** [comment]
I don't see why it needs to be rolled into any option like `stopSpecOnExpectationFailure`. If a test fails you should certainly be able to try running other tests. In most well written suites other tests may still pass. But if any part of the setup for a test fails, the output is meaningless. Is this in dispute?

**Comment by thw0rted** [comment]
I spent some time putting together a repro to show how async promise rejection in `beforeEach` wasn't causing specs to "fail" as the docs describe, but it turns out what I actually *meant* was "stop".  The spec does *fail* before `it` ever runs, as the docs promise, I just didn't know what to ask for.

There's similar vocabulary confusion in all these discussions.  For example, upthread @slackersoft says that a "failing expectation" can't always cause the current function to stop running.  But the problem here, and in several related issues, is that *rejecting* or *throwing* (depending if your `before`/`it` function is async) is not actually the same thing as a failing expectation, at least logically.  (Maybe it's treated exactly the same in the code, but that's an implementation detail.)

I would contend that, *out of the box* (by default, without changing a config file, etc), it does make sense to continue past `expect(true).toBeFalse()` but it does not make sense to run the `it` function when the `beforeEach` / `beforeAll` function throws or rejects.  I expect that case to play out exactly the same as how `throw "Oops"; expect(true).toBeFalse();` never executes the expectation.  `throw` / rejection provides a means of flow-control to the setup function, a strong signal that the test can't continue.  The current behavior ignores that signal.  Why introduce a new flow-control construct instead of just respecting the existing one?

**Comment by sgravrock** [comment]
> Why introduce a new flow-control construct instead of just respecting the existing one?

As is so often the case, there's a fair bit of complexity hiding behind the word "just'.

I think your interpretation makes sense for the simple case of a spec that only has a beforeEach associated with it. Things get more complicated when you consider afterEach. To stretch the flow control analogy a bit further: an afterEach is like a try/finally that wraps all of the preceding code including the spec, beforeEach functions, and any previous afterEach functions. It has to run even when the preceding functions fail or else state can leak from spec to spec. It's true that an afterEach often cleans up after a specific beforeEach, but that's a matter of programmer intent and not something that's encoded in the interface. So Jasmine has to run them all no matter what.

Given that the afterEach functions all have to run, the next question is what to do about beforeEach functions that come after a failure. Letting them run comes with some of the same drawbacks as letting the spec run. Skipping them increases the risk that a corresponding afterEach will misbehave because the system isn't in the state that it expected. Both choices have drawbacks but I think it's probably safer to run beforeEach functions that come after a failure than to skip them.

In other words, the general case isn't `[beforeEach, it]`. It's `[beforeEach, beforeEach, beforeEach, it, afterEach, afterEach]`. And the only thing in that second list that's definitely safe to skip after a failure is the `it`.

All of the above is potentially specific to beforeEach/afterEach. beforeAll/afterAll are somewhat different in both concept and implementation, and I haven't fully thought through the consequences of skipping them after an error occurs.

**Comment by thw0rted** [comment]
Good points. I hadn't really brought up any `after` functions.  You definitely want cleanup to run, though it's probably best to code the cleanup defensively (guards for undefined values, etc.) regardless.  I still think it's worth pursuing *some* way of canceling execution of whatever methods can be safely canceled.  

As for *how* to do that, I had a few ideas:

* before/it callbacks throw (reject) to request stopping
* before/it callbacks return a (Promise of) literal `false` to request stopping
* add `done.stop()` function, must take done-argument and stop explicitly

I do like the first one but it has the disadvantage of being a breaking change, while I think all the rest could be implemented in a mostly backwards compatible manner.  You could make it non-breaking by opting in to the new behavior with an options-object passed to before/it, e.g. `beforeEach("setup foo", {stopOnError: true}, () => ...)`.

**Comment by sgravrock** [comment]
> - before/it callbacks throw (reject) to request stopping

This is what I was thinking of. It's a breaking change, but not a big impactful one. I see no problem with including it in the next major release.

> - before/it callbacks return a (Promise of) literal false to request stopping

I think that's actually more of a breaking change than the first idea. Returning a promise resolved to literal false is silly but valid, and it's a good bet that somebody, somewhere, is doing it.

> - add done.stop() function, must take done-argument and stop explicitly

I don't want to add new functionality that's only available to callback-style async functions. Most people are better off using async/await or at least explicit promises these days and only falling back to callback-style async as a last resort.

**Comment by thw0rted** [comment]
Actually I thought about the second one (literal false) a bit and it's not even silly, it's probably pretty common.  Maybe I have a `load(): Promise<boolean>` that resolves with `true` if the thing was cached and `false` if it wasn't, but for this particular spec I just want to make sure some side-effect-observing expectations I set up get to run before the spec is finished, so I `return instance.load();`. I didn't care which value was resolved, just that the returned promise completed without rejecting.  It would be pretty bad for those tests to break just because I didn't read the release notes.

I think changing the throw/reject behavior is the right call.

**Comment by csslyion** [comment]
It is painful when you wake up and realize that the one you love most is hiding feelings or identity about him/her
